File: 399-EvaluateDivision/399-EvaluateDivision.cpp
// Last updated: 7/24/2025, 8:31:18 AM
class Solution {
public:
    vector<double> calcEquation(const vector<vector<string>>& equations, const vector<double>& values, const vector<vector<string>>& queries) {
        unordered_map<string, unordered_map<string, double>> adjList;
        vector<double> ans;

        for (int i = 0; i < equations.size(); i++) {
            adjList[equations[i][0]][equations[i][1]] = values[i];
            adjList[equations[i][1]][equations[i][0]] = 1.0 / values[i];
        }

        for (const vector<string>& query : queries) {
            double result = -1.0;

            if (adjList.find(query[0]) != adjList.end() && adjList.find(query[1]) != adjList.end()) {
                if (query[0] == query[1]) {
                    result = 1.0;
                } else {
                    unordered_map<string, bool> visited;
                    result = dfs(query[0], query[1], 1.0, adjList, visited);
                }
            }

            ans.push_back(result);
        }
        return ans;
    }

private:
    double dfs(const string& from, const string& to, double product, unordered_map<string, unordered_map<string, double>>& adjList, unordered_map<string, bool>& visited) {
        if (from == to) {
            return product;
        }

        visited[from] = true;
        
        for (const auto& neighbor : adjList[from]) {
            if (!visited[neighbor.first]) {
                double temp = dfs(neighbor.first, to, product * neighbor.second, adjList, visited);
                if (temp != -1.0) {
                    return temp;
                }
            }
        }

        return -1.0;
    }
};

File: 399-EvaluateDivision/399-EvaluateDivision.py
# Last updated: 7/24/2025, 8:31:13 AM
class Solution:
    def calcEquation(self, equations, values, queries):
        def find(node):
            if node != parents[node][0]:
                p, v = find(parents[node][0])
                parents[node] = (p, v * parents[node][1])
            return parents[node]

        def union(x, y, val):
            root_x, val_x = find(x)
            root_y, val_y = find(y)
            if root_x != root_y:
                parents[root_x] = (root_y, val_y / val_x * val)

        parents = {}
        results = []

        for (a, b), val in zip(equations, values):
            if a not in parents and b not in parents:
                parents[a] = (b, val)
                parents[b] = (b, 1.0)
            elif a not in parents:
                parents[a] = (b, val)
            elif b not in parents:
                parents[b] = (a, 1 / val)
            else:
                union(a, b, val)

        for start, end in queries:
            if start not in parents or end not in parents:
                results.append(-1.0)
                continue

            root_start, val_start = find(start)
            root_end, val_end = find(end)

            if root_start != root_end:
                results.append(-1.0)
            else:
                results.append(val_start / val_end)

        return results
